ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java

```diff
@@ -1039,7 +1039,7 @@ public static List<String> mergeUniqElems(List<String> src, List<String> dest) {
     return src;
   }
 
-  private static final String tmpPrefix = "_tmp.";
+  private static final String tmpPrefix = "-tmp.";
   private static final String taskTmpPrefix = "_task_tmp.";
```

Comment:
Just asking. Above the tempPrefix is -tmp here it is _task_tmp is this OK?

Reply:
I knew this comment would come :) 
The problem here is that there was an optimisation which avoids a copy of files from _tmp to -ext (HIVE-25990).
Unfortunately, the merge task was not configured after this optimisation hence it never invoked the merge task and it went on with the move task.

To make merge task work on _tmp directory is not possible since Hadoop does not allow split generation for hidden paths (like _tmp). Hence I could of only 2 cases - 
1. Make merge task work by changing this prefix.
2. Disable the optimisation when there is a merge task present.

I thought of making the first method work and hence I have done this change.

Reply:
Some ideas about the change:
- The "-tmp" output would be visible to user if HS2 fails to delete the directory(non-partition external table)
- Duplicated records might happen if the merge task reads all output files under "-tmp"

Reply:
For union insert case, we have changed the tmp directory recently to: <table-dir>/<staging-dir>/-ext-10000/_tmp.HIVE_UNION_SUBDIR_1/<partition-dir>/HIVE_UNION_SUBDIR_1,
`Utilities.toTempPath(root)` in `getMergeInputDirName` will return the path `<table-dir>/<staging-dir>/-tmp.-ext-10000` which could be empty.

Reply:
> The "-tmp" output would be visible to user if HS2 fails to delete the directory(non-partition external table)

@dengzhhu653 Are there any generic issues by making it visible to user if HS2 fails to delete the directory? Unfortunately, Merge task is not possible on hidden directories.

Reply:
Hi @SourabhBadhya, this might not be a problem, as the staging-dir is a hidden directory starts with "." in general.

Reply:
> Duplicated records might happen if the merge task reads all output files under "-tmp"
@dengzhhu653 There is deletion happening before finalising the list of files, which removes any unnecessary files present in "-tmp". Please refer here - 
https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java#L1473-L1474

I have also added a test case which involves doing CTAS operation with a union select operation.

Reply:
I'm wondering if there is a case: the `Utilities.java` finalising the files before all the Tez tasks finishes: though the job is marked as success, but some Tez attempts are slowly to terminate due to heavy GC, or load, etc. The `Utilities.java` has kept all valid files into `filesKept`, I would suggest reading them from the manifest when resolving the merge task.

For CTAS with union select operation, actually should be `union all`, only `union all` will change the temp directory to `-ext-10000/_tmp.HIVE_UNION_SUBDIR_1/dynamic_partition-dir/HIVE_UNION_SUBDIR_1`